### ISSUE TITLE
Some models are using UnsignedByte for their indices; not supported in DirectX

### ISSUE THREAD

**Issue by jessey-git** [KhronosGroup/glTF-Sample-Models#187]
From what I've been able to gather from my own viewer and from the docs is that single bytes are not supported for use as index buffers in DirectX:
https://docs.microsoft.com/en-us/windows/desktop/direct3ddxgi/format-support-for-direct3d-11-0-feature-level-hardware 
https://docs.microsoft.com/en-us/windows/desktop/direct3ddxgi/hardware-support-for-direct3d-12-1-formats

Models like the newly created MultiUVTest are problematic, as are some other smaller models in the repo.

The Khronos Blender exporter has a way to force an index type so perhaps it would be best to use UnsignedShort types to keep the sample models as widely usable as possible?

**Comment by javagl** [comment]
Some of the manually created sample models (I guess the `SimpleTriangle` etc.) use byte indices because that's the most compact form to represent the asset. When there are fewer than 256 indices, then `UNSIGNED_BYTE` is sufficient.

On the one hand, one could argue that this is a form of premature optimization: Exactly **iff** there are so few indices, the memory that is saved by using bytes instead of short are negligible (namely, one can only save 256 bytes of memory). 

On the other hand, unsigned byte indices are valid according to the specification. So I'd say that it's good to have at least *some* sample models that cover this case, and loaders/renderers have to be able to cope with this. In doubt, they have to perform a conversion (which also shouldn't be expensive for 256 indices).

**Comment by jessey-git** [comment]
Yes, this only matters when using such models as-is; with no manual re-buffering in the app to convert to different formats.

I'm kind of torn as this format is definitely allowed by the spec so it's fine. I was just wondering if the sample models should always be in an, ugh, lowest common denominator format.  I realize that's a bit of an unfortunate thing to ask :)

Thinking more, I suppose it should be entirely up to the app to do the conversion to support the spec, even for these canonical sample files.  I'll leave it to the project authors to close this out as necessary.

**Comment by bghgary** [comment]
If you want to test out all possible types for indices, try using these assets from the glTF-Asset-Generator: https://github.com/bghgary/glTF-Asset-Generator/tree/master/Output/Mesh_PrimitiveMode

**Comment by emackey** [comment]
@jessey-git Yeah, sadly it's too late to drop it from the 2.0 spec since there are plenty of models "in the wild" using it, including ones created by the official Blender exporter with the default options.  I think the better solution is to update the DirectX app to detect it and up-convert as needed.  Will only need to up-convert very tiny meshes, so, should be low-impact at runtime I would expect.

**Comment by jessey-git** [comment]
Yup, I would never suggest a spec change for such a thing.

This topic is more about: "If there's a choice on which format to use for the glTF-Sample-Models repo, which should be chosen? The one that exactly fits the requirements of the index count, the one directly consumable with the most APIs, or the one where the decision doesn't matter?"

I'm fine with any of those outcomes :)

**Comment by emackey** [comment]
Since we're leaving it in the spec, we want these sample models to have good coverage of all the different parts of the spec.  That way it's clear from the official samples if a viewer isn't able to see some subset of them, and bugs get filed against the viewer for having incomplete support.  That should get us the broadest compatibility across the ecosystem.

**Comment by emackey** [comment]
Re-opening this, as it seems there is new movement toward deprecating `UNSIGNED_BYTE` usage by indicies.

Sample 2.0 models using this are:

- AlphaBlendModeTest
- MultiUVTest
- OrientationTest
- SpecGlossVsMetalRough
- TextureCoordinateTest
- VertexColorTest

Most of these came out of the Blender exporter (including the labels floating above the water bottles in SpecGlossVsMetalRough) and would need to be re-exported.  One model (MultiUVTest) is an external contribution, we might have to ask for an update.
